### PR TITLE
test: include independent package tests in strict type-check

### LINKED ISSUE #2406: test(tsconfig): type-check tests in paths/git/isolation/providers/workflows (pass 1, no cross-package deps)
Implementation pass 1 of #2302. Independently shippable — **no dependency on any other package**.

Measured by a full survey (baseline `bun run type-check` verified green first, then `**/*.test.ts` removed from every `exclude`):

| package | errors | files | dominant cause |
| --- | --- | --- | --- |
| `paths` | 8 | 3 | trivial |
| `git` | 159 | 1 | one file, one pattern — branded types |
| `isolation` | 231 | 4 | 204 in `providers/worktree.test.ts`, branded types |
| `providers` | 251 | 9 | 155 in `claude/provider.test.ts`, mock signature erasure |
| `workflows` | 464 | 11 | 275 in `dag-executor.test.ts`; **moved here from #2407 — see below** |
| **total** | **1,113** | **28** | |

## Why these five, and not the six the survey first grouped

The survey's headline grouping ("adopt now for paths, cli, server, git, isolation, providers") is **not safe as stated**, and the tsconfigs show why:

```
cli      excludes: [..., '../core/src/**/*.test.ts', '../server/...', '../adapters/...']
server   excludes: [..., '../core/src/**/*.test.ts', '../adapters/src/**/*.test.ts']
adapters excludes: [..., '../core/src/**/*.test.ts']
```

Those `../X/src/**/*.test.ts` entries exist precisely because those packages **include other packages' sources**. Un-excluding tests in `cli` or `server` therefore pulls in `core`'s test files too, so neither can go green until `core` is clean. They belong in pass 2 with `core`, not here.

`paths`, `git`, `isolation`, `providers` have no such cross-package includes. They are genuinely independent.

### `workflows` moved here from #2407 (verified 2026-08-05)

#2407 grouped `workflows` under *"core must land first."* Its tsconfig says otherwise — `include: ["src/**/*", "src/defaults/text-imports.d.ts"]`, no cross-package reach at all. Verified `include` reach across the repo:

```
paths / git / isolation / providers   own src only
workflows                             own src only          <- independent
core                                  ../workflows          <- src/defaults/text-imports.d.ts only, a .d.ts
adapters                              ../core ../workflows
server                                ../adapters ../core ../workflows
cli                                   ../adapters ../core ../server ../workflows
```

Nothing downstream pulls `workflows`' **tests** — every reference into it is that one declaration file. So its errors can be fixed and landed without waiting on `core`'s 891.

That matters beyond tidiness: the typed-primitive series (#2486–#2492) lives entirely in `packages/workflows`, and the old grouping put it behind the largest tranche in the project for no reason.

Count refreshed on `dev` @ `ed4f6acd`: **464** (was 435 when #2407 was written), 275 in `dag-executor.test.ts`, then `executor.test.ts` 89, `utils/tool-formatter.test.ts` 51, `loader.test.ts` 25, rest <= 6. By kind: 120 x TS2493, 78 x TS2352, 47 x TS2345, 46 x TS2769, 41 x TS2741.

**One overlap worth planning around:** some TS2322 `not assignable to type 'DagNode'` errors on hand-built node literals are generated by the current `z.infer<...> & { other?: never }` union, which #2486 replaces with a discriminated union. Doing them separately means fixing the same fixtures twice — land `workflows` here, then #2486 immediately after, or fold them.

## Root causes (all mechanical)

- **branded types (~346 repo-wide, concentrated here)** — raw `'main'` passed where `BranchName`/`RepoPath` is expected. No new machinery needed: `toBranchName()` and `toRepoPath()` already exist at `packages/git/src/types.ts:11,17`.
- **mock signature erasure** — `mock(() => …)` erases the callee signature, so `.mock.calls[0][3]` types as `never` or an out-of-range tuple index. Fix is retyping the mock declaration.
- **over-narrow `as` casts on mock returns (TS2352)** — `as unknown as X`.

## Do NOT introduce `tsconfig.test.json`

Type-checking tests under looser settings reintroduces the exact gap #2302 exists to close, and a second CI step is a second thing to skip. Measured CI cost of folding tests into the main config is **zero**: 27.4s with tests included vs 28.5s baseline for the whole-repo parallel `type-check`.

`packages/web` has never had the exclusion — 48 test files, type-checked and clean today. The practice already works in this repo.

## Acceptance

- `**/*.test.ts` removed from `exclude` in `packages/{paths,git,isolation,providers}/tsconfig.json`
- `bun run type-check` green
- `bun run test` still green — do not "fix" a type error by changing what a test asserts
- No `@ts-expect-error` or `any` added to silence errors; if a genuine type problem is found in production code, fix it and say so

## Reference

Full survey data, per-file concentration, and the #2299 reproduction are in #2302 and its comments. Top 10 files account for 51% of all repo-wide errors, so the work is far less uniform than the raw count suggests.

### PR DESCRIPTION
**Pull request by Wirasm**
## Problem and outcome

The normal repository type-check excluded every test in `paths`, `git`, `isolation`, `providers`, and `workflows`, so production contract drift across 95 test files could pass the authoritative build gate.

- **Outcome:** All 95 tests now compile in their package's ordinary strict TypeScript project and therefore in `bun run type-check` and CI.
- **Invariant:** Existing runtime assertions and package test-process boundaries remain unchanged while test doubles and fixtures stay aligned with current production and SDK contracts.
- **Scope boundary:** This does not include the dependency-ordered `core`/`adapters`/`server`/`cli` pass from #2407 or the #2486-#2492 workflow primitive refactors.
- **Root cause:** Each package already included `src/**/*`, but its `tsconfig.json` explicitly subtracted `**/*.test.ts`; the root command consequently reported success without compiling those tests.

## Review guidance

- **Feedback requested:** Verify the fixture corrections accurately reflect current contracts without weakening the behavioral assertions.
- **Start here:** `packages/workflows/tsconfig.json:6` — representative load-bearing change that moves tests into the existing strict project.
- **Review order:** The five `tsconfig.json` changes, branded git/isolation fixtures, provider SDK-shaped mocks, then workflow store/run/node fixtures.
- **Post-#2604 integration:** This branch is rebased onto `c41ed69d`. The four overlapping workflow test files and their production schemas/helpers were audited; #2604's composition behavior and assertions are preserved. Its newly added fixtures now implement the current async provider, executable-workflow, config-loader, and store contracts.
- **Post-#2603 integration:** This branch is now rebased onto `3634d158` after #2615/#2620. There is no direct file overlap and all five feature commits remain patch-equivalent. The expanded paths/dry-run tests and #2603 inventory guard are part of the compatibility gate.
- **Independent-review correction:** Both executor summary tests now fail unless the workflow result is successful and non-paused before asserting the summary, preserving the original fail-loud behavior across the current result union.
- **Lower-attention areas:** Repeated branded constructors, complete capability objects, and async-generator annotations are mechanical corrections validated by the unchanged runtime suites.
- **Known risk or uncertainty:** None.

## Solution

Each target package now excludes only dependencies and build output, allowing its existing `src/**/*` include to own production and test code together. The newly exposed diagnostics were resolved at test boundaries by using existing branded constructors, exact interface/SDK mock signatures, complete schema-derived fixtures, and explicit discriminated-union narrowing. No parallel test config, production type redesign, `any`, or `@ts-expect-error` escape hatch was added.

## Validation

- Target inventory vs `tsc --listFilesOnly` — 95/95 included, 0 missing: paths 9/9, git 2/2, isolation 12/12, providers 34/34, workflows 38/38.
- Each target package's existing `type-check` and `test` scripts — exit 0.
- Focused post-#2604 workflow type-check and composition/schema test suites — exit 0.
- `bun run validate` after rebasing onto #2604 — exit 0 as the authoritative composed pre-PR gate, including generated-artifact checks, all workspace and script type-checks, lint with zero warnings, formatting, installer tests, and all repository runtime suites.
- Original/current target set comparison — all original 95 test files remain and current inventory is 95; every file appears in its ordinary package compiler list.
- Expanded new-base checks — strict paths/workflows type-checks passed; `archon-paths.test.ts` passed 130 tests, `dry-run.test.ts` passed 39 tests, and #2603's inventory guard passed without selector drift.
- `bun run validate` after rebasing onto `3634d158` — exit 0, including the inventory guard in the normal root script-test batch.
- R10 focused correction checks — workflow strict type-check passed and all 84 `executor.test.ts` tests passed.
- `bun run validate` after the R10 correction — exit 0 at `03ffc80b`.
- **In progress:** Exact-head GitHub checks and a fresh full independent review at `03ffc80b`.

## Delivery considerations

| Concern | Impact and required action | Evidence |
|---|---|---|
| Compatibility / migration | Build-time enforcement expands; no runtime or data migration exists. | Existing package and repository runtime suites pass after rebasing onto #2604 and #2603/#2615/#2620. |
| Rollout / rollback | Package-local and reversible by restoring the exclusions and fixture-only corrections. | No production source file changed. |
| Observability | Future test/contract drift fails the normal root type-check and CI rather than a parallel optional check. | All 95 tests appear in ordinary package compiler lists. |

## Links

- Closes #2406
- Related #2407
- [Implementation plan](https://github.com/coleam00/Archon/issues/2406#issuecomment-5326489011)

## Summary by CodeRabbit

- **Tests**
  - Strengthened type validation across Git, isolation, workflows, and AI provider test suites.
  - Added coverage for workflow retries, cancellation, provider events, artifact handling, and lifecycle scenarios.
  - Updated fixtures and assertions to reflect current workflow and provider behavior.
  - Test files are now included in TypeScript compilation to catch compatibility issues earlier.
- **Refactor**
  - Cleaned up unused imports and improved test mock consistency without changing runtime behavior.

### CODE REVIEW

File: packages/workflows/src/condition-evaluator.test.ts [hunk 1]
```diff
@@ -2,7 +2,7 @@ import { describe, it, expect, mock } from 'bun:test';
 
 // --- Mock logger (MUST come before imports of modules under test) ---
 
-const mockLogFn = mock(() => {});
+const mockLogFn = mock((_data: unknown, _message?: string) => {});
```
[the remark this answers is not in the document]
Comment (Wirasm) [reply packages/workflows/src/condition-evaluator.test.ts]: Fixed in 4ed90c80. Added explicit return types to every cited callback, including logger, filesystem, resolver, Pi prompt, and telemetry fetch mocks. The final repository validation passes (`bun run validate`, exit 0).

**Comment by Wirasm** [comment]
---
pr: 2605
base: dev
head: test/2406-typecheck-independent-tests
reviewed: 2026-08-19T08:14:28Z
reviewed_head: 97272927347acbf4a0604664448d40d668d7d445
verdict: NEEDS FIXES
open_findings: 1
scopes: [code, seams, tests, types]
publication: https://github.com/coleam00/Archon/pull/2605#issuecomment-5327226097
---



# PR Review: #2605 — test: include independent package tests in strict type-check

## Signal

Needs one test-only correction. All 95 original and current target tests compile through the five ordinary strict package projects, #2603's unchanged inventory guard passes, the post-#2604 fixture fixes remain intact, `bun run validate` passes, and every GitHub check is green. However, three independent review scopes found that the two summary-propagation tests now accept the supported paused-success result and silently skip their original summary assertions, violating this PR's explicit assertion-preservation invariant.

- **Blocking:** 1 — `R10`
- **Non-blocking:** 0 open
- **Resolved since previous review:** 0; all prior fixed rows remain fixed, while `R3` is no longer tracked against the completed #2401 workstream
- **Tracked follow-ups:** None
- **Validation:** PASS — 95/95 compiler inclusion, focused post-#2603 compatibility tests, final-head `bun run validate`, and all GitHub checks

## Findings

| ID | Severity | Finding | State |
|---|---|---|---|
| `R1` | Important | Codex test capabilities contradicted the canonical provider definition | FIXED |
| `R2` | Important | Pi session and registry mocks duplicated SDK-owned contracts | FIXED |
| `R3` | Suggestion | Compiler inclusion has no committed inventory guard | DECLINED |
| `R4` | Suggestion | Changed callbacks lacked repository-required explicit return types | FIXED |
| `R5` | Suggestion | Automated review classified typed Codex fixtures as new behavioral coverage | NOT A FINDING |
| `R6` | Suggestion | Automated review requested generic docstrings for test helpers | DECLINED |
| `R7` | Important | Malformed persisted usage can suppress the required completion event undetected | FIXED |
| `R8` | Important | Missing switched-provider node can satisfy the model-clearing assertion | FIXED |
| `R9` | Important | Missing loop-group body can satisfy the persist-session assertion | FIXED |
| `R10` | Important | Paused results can bypass both summary-propagation assertions | OPEN |

## Detailed Findings

<details>
<summary><code>R1</code> — Codex test capabilities contradicted the canonical provider definition</summary>

**Impact:** The test provider advertised native tools and skills that Codex does not support, allowing capability-dependent tests to exercise an impossible Codex path.

**Evidence:** `packages/workflows/src/dag-executor.test.ts:210-211` delegates to `getProviderCapabilities('codex')`; the authoritative values remain `skills: false` and `nativeTools: false` in `packages/providers/src/codex/capabilities.ts:10` and `:25`.

**Required outcome:** Carry the canonical registered Codex capabilities instead of reconstructing them in the fixture.

**Found by:** `code-reviewer`, `seam-analyzer`

**Disposition:** FIXED; the canonical registry value remains the source of truth at `97272927`, independently verified by the fresh code, seam, and type reviews.

</details>

<details>
<summary><code>R2</code> — Pi session and registry mocks duplicated SDK-owned contracts</summary>

**Impact:** Handwritten mock parameters and results could continue compiling when the Pi SDK changed, leaving the newly strict test project unable to detect contract drift at the SDK boundary.

**Evidence:** `packages/providers/src/community/pi/provider.test.ts:74-140` types session and registry mocks from the SDK, while `:2506-2537` derives extension registration and session-option shapes from SDK exports.

**Required outcome:** Derive callable and result shapes from imported Pi SDK exports, keeping assertions only at deliberately fabricated object boundaries.

**Found by:** `code-reviewer`, `seam-analyzer`

**Disposition:** FIXED; the SDK-derived mocks and complete model fixtures remain present at `97272927`, verified by the fresh code, seam, and type reviews.

</details>

<details>
<summary><code>R3</code> — Compiler inclusion has no committed inventory guard</summary>

**Impact:** A future config could re-add a test exclusion without a dedicated compiler-root inventory check detecting it.

**Evidence:** The five ordinary projects currently include all 95 target tests with zero missing or extra. The merged #2603 guard at `scripts/test-inventory.test.ts:94-127` checks runtime package selectors, not TypeScript compiler roots, and PR #2605 does not alter that guard or any package test selector.

**Required outcome:** No correction is required in this PR. Its accepted invariant is exact compiler inclusion at the reviewed head, which is directly proved; a permanent compiler meta-guard was not part of #2401 or this plan.

**Found by:** `pr-test-analyzer`

**Disposition:** DECLINED. The prior review tracked this suggestion against #2401, but that issue is now completed by #2603's intentionally runtime-focused selector guard. Fresh test and type review confirmed that compiler inclusion remains 95/95 and that no committed compiler guard is required for this PR's outcome.

</details>

<details>
<summary><code>R4</code> — Changed callbacks lacked explicit return types</summary>

**Impact:** Changed test callbacks violated the repository's complete-type-annotation rule even though inference was correct.

**Evidence:** The [CodeRabbit review thread](https://github.com/coleam00/Archon/pull/2605#discussion_r3803275148) identified logger, filesystem, resolver, Pi prompt, and telemetry fetch callbacks. Their explicit `void`, `Promise<void>`, `Promise<Response>`, and `Promise<IsolatedEnvironment>` annotations remain in the final diff.

**Required outcome:** Keep explicit return types on every cited changed callback.

**Found by:** `CodeRabbit`

**Disposition:** FIXED; fresh code and type reviews verified the annotations at `97272927`, and validation passes.

</details>

<details>
<summary><code>R5</code> — Automated review classified typed Codex fixtures as new behavioral coverage</summary>

**Impact:** None; treating the warning as actionable would remove existing assertions that merely needed sound mock signatures.

**Evidence:** The PR types existing Codex cases and SDK mock surfaces; it does not introduce new Codex behavior. The unchanged runtime suite passes after the type corrections.

**Required outcome:** No correction required.

**Found by:** `CodeRabbit summary check`

**Disposition:** NOT A FINDING; the warning conflates type-correction edits with newly introduced behavioral scope.

</details>

<details>
<summary><code>R6</code> — Automated review requested generic docstrings for test helpers</summary>

**Impact:** None; generic docstrings on local test callbacks would add unrelated prose without improving the strict compiler invariant.

**Evidence:** Repository guidance requires complete type annotations but does not impose the external summary check's 80% docstring threshold on test helpers. The changed helpers are local and their contracts are expressed by imported types.

**Required outcome:** No correction required.

**Found by:** `CodeRabbit summary check`

**Disposition:** DECLINED as an unrelated generic heuristic that conflicts with the PR's minimal fixture-only scope.

</details>

<details>
<summary><code>R7</code> — Malformed persisted usage can suppress the required completion event undetected</summary>

**Impact:** The malformed-token resume test could pass if the executor stopped writing the required `node_completed` audit event.

**Evidence:** `packages/workflows/src/dag-executor.test.ts:8738-8742` filters matching completion events, requires exactly one, and then reads its data non-optionally before checking that `tokens` is absent.

**Required outcome:** Require exactly one completion event before inspecting its data.

**Found by:** `seam-analyzer`, `pr-test-analyzer`

**Disposition:** FIXED; all four fresh specialists verified that the fail-loud assertion remains present at `97272927`.

</details>

<details>
<summary><code>R8</code> — Missing switched-provider node can satisfy the model-clearing assertion</summary>

**Impact:** The include-expander test could pass if composition dropped the switched-provider node entirely.

**Evidence:** `packages/workflows/src/include-expander.test.ts:1662-1669` requires exactly one expanded node before reading its `model`.

**Required outcome:** Require the expanded node before checking that its model is cleared.

**Found by:** `pr-test-analyzer`

**Disposition:** FIXED; the fresh code, seam, test, and type reviews verified the existence assertion at `97272927`.

</details>

<details>
<summary><code>R9</code> — Missing loop-group body can satisfy the persist-session assertion</summary>

**Impact:** The loader test could pass if composition retained the loop-group shell but deleted its body node.

**Evidence:** `packages/workflows/src/loader.test.ts:5247-5253` requires the loop-group configuration and body node before inspecting `persist_session`.

**Required outcome:** Require the body node before checking that persistence is absent.

**Found by:** `pr-test-analyzer`

**Disposition:** FIXED; the fresh code, seam, test, and type reviews verified the body assertion at `97272927`.

</details>

<details>
<summary><code>R10</code> — Paused results can bypass both summary-propagation assertions</summary>

**Impact:** A regression that returns the supported paused-success result for a completed workflow leaves both summary tests green without checking a completed result or its summary. Downstream, the CLI would print “Workflow paused,” while the orchestrator would suppress the completion result card and summary.

**Evidence:** `packages/workflows/src/executor.test.ts:1279-1282` and `:1298-1301` assert only `success: true`, then conditionally run the summary assertion when `paused` is absent. The union explicitly permits `{ success: true; paused: true; ... }` at `packages/workflows/src/schemas/workflow.ts:316-319`, production returns it for a paused persisted status at `packages/workflows/src/executor.ts:1951-1954`, and consumers branch on it at `packages/core/src/orchestrator/orchestrator.ts:501-504` and `packages/cli/src/commands/workflow.ts:2115-2120`. Code, seam, and test reviewers independently reproduced the skipped-assertion path.

**Required outcome:** In both summary cases, fail unless the result is successful and non-paused, then assert the expected summary. Keep the correction test-only; the paused-success union is intentional and does not need redesign.

**Found by:** `code-reviewer`, `seam-analyzer`, `pr-test-analyzer`

**Disposition:** OPEN at `97272927`.

</details>

## Agent Coverage

| Scope | Result |
|---|---|
| code | `R10`; prior `R1`, `R2`, `R4`, and `R7`-`R9` verified FIXED |
| seams | `R10`; SDK, branded, persisted/resumed, validator, and compiler-inclusion counterparts otherwise aligned |
| tests | `R10`; `R7`-`R9` verified FIXED, and `R3` re-evaluated after #2603 |
| types | No type findings; confirmed `R10` is a test-only fail-open narrowing gap, not a production union defect, and verified prior type-related fixes |

## Validation

| Command | Result | Evidence |
|---|---|---|
| `bun run validate` | PASS | Exit 0 at `97272927`; generated artifacts, all workspace and script type-checks, zero-warning lint, formatting, installer tests, isolated package suites, and root script tests passed. |
| Target inventory vs `bun x tsc --noEmit --listFilesOnly -p packages/<pkg>/tsconfig.json` | PASS | paths 9/9, git 2/2, isolation 12/12, providers 34/34, workflows 38/38; original and current inventories are both 95, with zero missing or extra. |
| `bun test src/archon-paths.test.ts` | PASS | 130 passed, 0 failed. |
| `bun test src/dry-run.test.ts` | PASS | 39 passed, 0 failed. |
| `bun test ./scripts/test-inventory.test.ts` | PASS | #2603's unchanged package-selector guard passed, 1 test and 0 failures. |
| `git diff --check 3634d158...97272927` plus scope/escape audit | PASS | No whitespace errors, production changes, selector/guard changes, `any`, TypeScript suppression, or parallel/looser test config. |
| `gh pr checks 2605` | PASS | Ubuntu, Windows, Docker build, schema-upgrade, and CodeRabbit are green at `97272927`. |
| Local/remote/PR identity verification | PASS | Local HEAD, remote branch, and PR head are `97272927347acbf4a0604664448d40d668d7d445`; PR base is `3634d158c1c80f8a5b6f9d4c2a2ee0657d7d1b19`; worktree clean. |

## Verdict

**NEEDS FIXES**

Correct `R10` by making both summary tests fail closed on failure or pause, rerun the focused workflow tests and authoritative validation, then perform a fresh exact-head re-review. Do not redesign the production result union and do not merge this head.
